File: ClassicalSharp/Rendering/ChunkSorter.cs
// ClassicalSharp copyright 2014-2016 UnknownShadow200 | Licensed under MIT
using System;
using ClassicalSharp.GraphicsAPI;
using OpenTK;

namespace ClassicalSharp.Renderers {
	
	public static class ChunkSorter {

		public static void UpdateSortOrder( Game game, ChunkUpdater updater ) {
			Vector3 cameraPos = game.CurrentCameraPos;
			Vector3I newChunkPos = Vector3I.Floor( cameraPos );
			newChunkPos.X = (newChunkPos.X & ~0x0F) + 8;
			newChunkPos.Y = (newChunkPos.Y & ~0x0F) + 8;
			newChunkPos.Z = (newChunkPos.Z & ~0x0F) + 8;			
			if( newChunkPos == updater.chunkPos ) return;
			
			ChunkInfo[] chunks = game.MapRenderer.chunks;
			int[] distances = updater.distances;		
			updater.chunkPos = newChunkPos;
			
			for( int i = 0; i < distances.Length; i++ ) {
				ChunkInfo info = chunks[i];
				distances[i] = Utils.DistanceSquared( info.CentreX, info.CentreY, info.CentreZ, 
				                                     updater.chunkPos.X, updater.chunkPos.Y, updater.chunkPos.Z );
			}

			// NOTE: Over 5x faster compared to normal comparison of IComparer<ChunkInfo>.Compare
			if( distances.Length > 1 )
				QuickSort( distances, chunks, 0, chunks.Length - 1 );
			
			Vector3I pPos = newChunkPos;
			for( int i = 0; i < chunks.Length; i++ ) {
				ChunkInfo info = chunks[i];
				int dX1 = (info.CentreX - 8) - pPos.X, dX2 = (info.CentreX + 8) - pPos.X;
				int dY1 = (info.CentreY - 8) - pPos.Y, dY2 = (info.CentreY + 8) - pPos.Y;
				int dZ1 = (info.CentreZ - 8) - pPos.Z, dZ2 = (info.CentreZ + 8) - pPos.Z;
				
				// Back face culling: make sure that the chunk is definitely entirely back facing.
				info.DrawLeft = !(dX1 <= 0 && dX2 <= 0);
				info.DrawRight = !(dX1 >= 0 && dX2 >= 0);
				info.DrawFront = !(dZ1 <= 0 && dZ2 <= 0);
				info.DrawBack = !(dZ1 >= 0 && dZ2 >= 0);
				info.DrawBottom = !(dY1 <= 0 && dY2 <= 0);
				info.DrawTop = !(dY1 >= 0 && dY2 >= 0);
			}
			updater.RecalcBooleans( false );
			//SimpleOcclusionCulling();
		}
		
		static void QuickSort( int[] keys, ChunkInfo[] values, int left, int right ) {
			while( left < right ) {
				int i = left, j = right;
				int pivot = keys[(i + j) / 2];
				// partition the list
				while( i <= j ) {
					while( pivot > keys[i] ) i++;
					while( pivot < keys[j] ) j--;
					
					if( i <= j ) {
						int key = keys[i]; keys[i] = keys[j]; keys[j] = key;
						ChunkInfo value = values[i]; values[i] = values[j]; values[j] = value;
						i++; j--;
					}
				}
				
				// recurse into the smaller subset
				if( j - left <= right - i ) {
					if( left < j )
						QuickSort( keys, values, left, j );
					left = i;
				} else {
					if( i < right )
						QuickSort( keys, values, i, right );
					right = j;
				}
			}
		}
	}
}

File: src/Client/ChunkSorter.c
#include "ChunkSorter.h"
#include "ChunkUpdater.h"
#include "MapRenderer.h"
#include "GameProps.h"
#include "Vectors.h"
#include "Vector3I.h"
#include "Constants.h"

void ChunkSorter_UpdateSortOrder(void) {
	Vector3 cameraPos = Game_CurrentCameraPos;
	Vector3I newChunkPos;
	Vector3I_Floor(&newChunkPos, &cameraPos);

	newChunkPos.X = (newChunkPos.X & ~CHUNK_MAX) + HALF_CHUNK_SIZE;
	newChunkPos.Y = (newChunkPos.Y & ~CHUNK_MAX) + HALF_CHUNK_SIZE;
	newChunkPos.Z = (newChunkPos.Z & ~CHUNK_MAX) + HALF_CHUNK_SIZE;
	/* Same chunk, therefore don't need to recalculate sort order. */
	if (Vector3I_Equals(&newChunkPos, &ChunkUpdater_ChunkPos)) return;

	Vector3I pPos = newChunkPos;
	ChunkUpdater_ChunkPos = pPos;
	if (MapRenderer_ChunksCount == 0) return;

	Int32 i = 0;
	for (i = 0; i < MapRenderer_ChunksCount; i++) {
		ChunkInfo* info = MapRenderer_SortedChunks[i];

		/* Calculate distance to chunk centre*/
		Int32 dx = info->CentreX - pPos.X, dy = info->CentreY - pPos.Y, dz = info->CentreZ - pPos.Z;
		ChunkUpdater_Distances[i] = dx * dx + dy * dy + dz * dz; /* TODO: do we need to cast to unsigned for the mulitplies? */

		/* Can work out distance to chunk faces as offset from distance to chunk centre on each axis. */
		Int32 dXMin = dx - HALF_CHUNK_SIZE, dXMax = dx + HALF_CHUNK_SIZE;
		Int32 dYMin = dy - HALF_CHUNK_SIZE, dYMax = dy + HALF_CHUNK_SIZE;
		Int32 dZMin = dz - HALF_CHUNK_SIZE, dZMax = dz + HALF_CHUNK_SIZE;

		/* Back face culling: make sure that the chunk is definitely entirely back facing. */
		info->DrawXMin = !(dXMin <= 0 && dXMax <= 0);
		info->DrawXMax = !(dXMin >= 0 && dXMax >= 0);
		info->DrawZMin = !(dZMin <= 0 && dZMax <= 0);
		info->DrawZMax = !(dZMin >= 0 && dZMax >= 0);
		info->DrawYMin = !(dYMin <= 0 && dYMax <= 0);
		info->DrawYMax = !(dYMin >= 0 && dYMax >= 0);
	}

	ChunkSorter_QuickSort(0, MapRenderer_ChunksCount - 1);
	ChunkUpdater_ResetPartFlags();
	//SimpleOcclusionCulling();
}

void ChunkSorter_QuickSort(Int32 left, Int32 right) {
	ChunkInfo** values = MapRenderer_SortedChunks;
	Int32* keys = ChunkUpdater_Distances;
	while (left < right) {
		Int32 i = left, j = right;
		Int32 pivot = keys[(i + j) / 2];
		/* partition the list */
		while (i <= j) {
			while (pivot > keys[i]) i++;
			while (pivot < keys[j]) j--;

			if (i <= j) {
				Int32 key = keys[i]; keys[i] = keys[j]; keys[j] = key;
				ChunkInfo* value = values[i]; values[i] = values[j]; values[j] = value;
				i++; j--;
			}
		}

		/* recurse into the smaller subset */
		if (j - left <= right - i) {
			if (left < j) ChunkSorter_QuickSort(left, j);
			left = i;
		} else {
			if (i < right) ChunkSorter_QuickSort(i, right);
			right = j;
		}
	}
}